# -------DATA CLEANING-------

## -------DATA INSPECTION-------

In [1]:
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
df = pd.read_csv("messy_customer_sales_data.csv")
df

,Customer_ID,Name,Gender,Age,City,Signup_Date,Last_Purchase_Date,Purchase_Amount,Feedback_Score,Email,Phone_Number,Country
0,CUST4371,Paul Wilson,m,52.0,KOLKATA,2025-06-26,2025-05-17,26944.0,1.0,peckvictoria@example.com,2131107701,India
1,CUST5957,Jason Thomas,M,51.0 years,NaN,2021-02-17,2025-07-22,44152.0,2.0,owensanthony@example.com,1080761560,india
2,CUST3754,Brittney Martinez,F,62.0,hyderabad,2023-11-05,2024-12-08,31745.0,2.0,tara39@example.org,8981006345,India
3,CUST2934,Brenda Pierce,FEMALE,40.0,hyderabad,2022-03-13,2025-10-02,39674.0,1.0,berrynancy@example.org,8228064204,india
4,CUST5683,Matthew Carroll,f,41.0,CHENNAI,2024-04-05,2024-12-15,NaN,8.0,denise84@example.org,2665569480,India
...,...,...,...,...,...,...,...,...,...,...,...,...
10195,CUST10767,Robert Lewis,female,35.0 years,delhi,2020-12-08,2025-01-25,24167.0,9.0,barrycrane@example.net,5004696571,InDia
10196,NaN,Diane Evans,M,53.0,bangalore,2023-12-31,2025-05-07,11639.0,7.0,lisadennis@example.net,5200349941,IND
10197,CUST6315,Joshua Martinez,m,25.0,hyderabad,2022-02-15,2025-01-11,43832.0,2.0,kelli74@example.org,8147428496,India
10198,CUST4812,Sarah Miller,FEMALE,55.0,NaN,2021-03-16,2025-05-14,18827.0,10.0,dawn84@example.org,2987564247,InDia


In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10200 entries, 0 to 10199
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Customer_ID         9177 non-null   object 
 1   Name                10200 non-null  object 
 2   Gender              9174 non-null   object 
 3   Age                 9249 non-null   object 
 4   City                9184 non-null   object 
 5   Signup_Date         10200 non-null  object 
 6   Last_Purchase_Date  9188 non-null   object 
 7   Purchase_Amount     9179 non-null   float64
 8   Feedback_Score      9177 non-null   float64
 9   Email               10200 non-null  object 
 10  Phone_Number        10200 non-null  int64  
 11  Country             9468 non-null   object 
dtypes: float64(2), int64(1), object(9)
memory usage: 956.4+ KB


### Problem inspected from above are:
      1. Datatypes
        - Age should have integer Dtype
        - Signup_Date should have date & time Dtype
        - Last_Purchase_Date should have date & time Dtype 

In [6]:
df.describe().round() #Summary Statistics Techniques

,Purchase_Amount,Feedback_Score,Phone_Number
count,9179.0,9177.0,1.020000e+04
mean,29090.0,5.0,4.979974e+09
std,208697.0,3.0,2.902593e+09
min,-500.0,1.0,9.208990e+05
25%,12295.0,3.0,2.449157e+09
50%,24330.0,5.0,4.988639e+09
75%,37130.0,8.0,7.510448e+09
max,9999999.0,10.0,9.994402e+09


### Problems inspected from above are:
      1. Outliers
        - in purchase amount, it can be seen the max is more above than mean value. So, it is outlier

In [7]:
df.isnull().sum() #Check for missing values

Customer_ID           1023
Name                     0
Gender                1026
Age                    951
City                  1016
Signup_Date              0
Last_Purchase_Date    1012
Purchase_Amount       1021
Feedback_Score        1023
Email                    0
Phone_Number             0
Country                732
dtype: int64

In [9]:
df.shape

(10200, 12)

### Problems inspected from above are:
      1. Null Values from 10200 records
      
          Customer_ID           1023
          Name                     0
          Gender                1026
          Age                    951
          City                  1016
          Signup_Date              0
          Last_Purchase_Date    1012
          Purchase_Amount       1021
          Feedback_Score        1023
          Email                    0
          Phone_Number             0
          Country                732

In [11]:
df.isnull().sum()/len(df)*100 #Percentage of missing values

Customer_ID           10.029412
Name                   0.000000
Gender                10.058824
Age                    9.323529
City                   9.960784
Signup_Date            0.000000
Last_Purchase_Date     9.921569
Purchase_Amount       10.009804
Feedback_Score        10.029412
Email                  0.000000
Phone_Number           0.000000
Country                7.176471
dtype: float64

In [13]:
df.dropna().shape

(4528, 12)

### Note: We have more missing values in several columns. We can drop the columns but dropping the columns may leas loss of numbers of records. We can shee above after dropping we have 4528 records i.e 60% of data we have lost. So, directly dropping the columns is not a good idea.

In [ ]:
df.duplicates

(10200, 12)

In [15]:
df[df.duplicated()].shape

(15, 12)

In [16]:
df[df.duplicated()] #Check for duplicate records

,Customer_ID,Name,Gender,Age,City,Signup_Date,Last_Purchase_Date,Purchase_Amount,Feedback_Score,Email,Phone_Number,Country
2032,CUST2755,Casey Campbell,male,NaN,Hyderabad,2023-12-01,2024-11-04,11778.0,3.0,joseph49@example.com,3844741384,India
3191,NaN,Travis Schneider,f,51.0,NaN,2025-04-20,2025-02-28,48750.0,5.0,bianca77@example.net,7623118785,India
4312,CUST8305,Stanley Cain,FEMALE,51.0,Bangalore,2023-08-14,2024-11-01,11432.0,6.0,calderonbrenda@example.org,4889366037,India
5611,CUST6288,Jonathon Kim,male,46.0,Chennai,2023-11-22,2025-03-21,3301.0,6.0,tparker@example.com,5266662560,India
5680,CUST2695,Lisa Durham,NaN,46.0,Chennai,2021-04-18,2024-10-10,6511.0,3.0,andrew34@example.com,2235480149,India
5899,CUST8841,Amanda Hill,NaN,34.0,chennai,2022-09-11,2025-03-13,42092.0,4.0,ryanriley@example.net,3466026146,India
6070,CUST10780,Emily Smith,NaN,23.0,Kolkata,2023-12-09,2024-10-28,18027.0,5.0,perezchristopher@example.com,3032282870,India
6240,CUST9745,David Morales,NaN,56.0,NaN,2023-02-18,2024-12-14,49107.0,7.0,juliaknapp@example.com,9436163110,India
6425,CUST4631,George Villa,m,26.0,Mumbai,2022-05-21,2025-06-17,NaN,8.0,davidmarquez@example.com,8869465862,India
6498,CUST1711,Stephanie Elliott,NaN,44.0,MUMBAI,2023-05-17,2025-07-12,11349.0,9.0,nfischer@example.com,3270872546,NaN


## Problems inspected from above are:
  1. Duplicate records
    - There are 15 rows and 12 columns with duplicates value

In [20]:
df['Customer_ID'].value_counts() #Check for duplicate records in a specific column

Customer_ID
CUST3374     2
CUST7749     2
CUST10800    2
CUST6588     2
CUST9062     2
            ..
CUST10204    1
CUST4647     1
CUST9548     1
CUST9966     1
CUST4812     1
Name: count, Length: 9000, dtype: int64

### Since, Customer_ID is unique, it's duplicate value show duplicate record(row). And using the function value_counts that gives frequency of data, duplicate records have generated above

In [21]:
df[df['Customer_ID'] == 'CUST3374']

,Customer_ID,Name,Gender,Age,City,Signup_Date,Last_Purchase_Date,Purchase_Amount,Feedback_Score,Email,Phone_Number,Country
2914,CUST3374,Timothy Wallace,m,58.0,Bangalore,2024-03-08,2025-08-21,45779.0,1.0,glenn95@example.org,9687701793,India
3047,CUST3374,Timothy Wallace,m,58.0,bangalore,2024-03-08,2025-08-21,45779.0,1.0,glenn95@example.org,9687701793,India


### Note: There were 15 duplicate rows/records, but these rows were not detected initially because the records had inconsistent formatting. For example, rows 2914 and 3047 contain the same records, but the first letter in the City column has a different case. The earlier function detected these as different records according to the letter case in city column, but it got identified as duplicates when frequency counts were applied. These row has more inconsistencies.

In [22]:
for col in df.columns:
    if df[col].nunique() < 20:
        print(df[col].value_counts())
        print("-"*50)

Gender
f         1184
M         1171
m         1163
F         1157
MALE      1131
female    1128
male      1121
FEMALE    1119
Name: count, dtype: int64
--------------------------------------------------
City
Kolkata       820
Mumbai        812
Chennai       784
Bangalore     773
Hyderabad     770
Delhi         763
CHENNAI       404
KOLKATA       395
MUMBAI        393
hyderabad     384
bangalore     383
DELHI         378
delhi         369
BANGALORE     363
HYDERABAD     360
mumbai        352
chennai       343
kolkata       338
Name: count, dtype: int64
--------------------------------------------------
Feedback_Score
2.0     952
4.0     947
7.0     938
6.0     927
3.0     913
8.0     912
1.0     907
9.0     903
10.0    901
5.0     877
Name: count, dtype: int64
--------------------------------------------------
Country
India    7132
IND       793
india     772
InDia     771
Name: count, dtype: int64
--------------------------------------------------


### Data Inspection from above code
  ### The above output shows the inconsistencies in categorical data present in gender, city and country columns.

## ----------DATA CLEANING----------

## -------Handling Missing Data------

In [25]:
df.dropna(subset = ['Customer_ID'], inplace = True)
df.isnull().sum()
# This drops all row with missing values in the 'Customer_ID' column.
# Because 'Customer_ID' is a unique identifier, we cannot have missing values in this column.

Customer_ID             0
Name                    0
Gender                934
Age                   859
City                  918
Signup_Date             0
Last_Purchase_Date    914
Purchase_Amount       927
Feedback_Score        905
Email                   0
Phone_Number            0
Country               664
dtype: int64

In [27]:
df['Age'].apply(lambda x: re.findall('[0-9]+', str(x)))

0        [52, 0]
1        [51, 0]
2        [62, 0]
3        [40, 0]
4        [41, 0]
          ...   
10194    [40, 0]
10195    [35, 0]
10197    [25, 0]
10198    [55, 0]
10199    [34, 0]
Name: Age, Length: 9177, dtype: object

In [28]:
# import re
# for i in df['Age'].apply(lambda x: re.findall('[0-9]+', str(x))):
#     if len(i) < 2:
#         print(i)

In [ ]:
import re

def extract_age(age):
    age_num = re.findall('[0-9]+', str(age))
    if len(age_num) > 0:
        return int(age_num[0])
    else:
        return age
df['age'] =df['Age'].apply(lambda x: extract_age(x))

0        52
1        51
2        62
3        40
4        41
         ..
10194    40
10195    35
10197    25
10198    55
10199    34
Name: Age, Length: 9177, dtype: object

### Let's decode the above code:

The value of each row in the Age column is passed to the lambda function. The lambda function then calls the extract_age() function, passing x (i.e., each value from the Age column) as an argument.

Inside the extract_age() function, the re.findall('[0-9]+', str(age)) function searches for numeric values. [0-9] matches any digit from 0 to 9, while + allows it to match one or more consecutive digits. For example, if the value is 23-58, findall() returns ['23', '58'].

In the if condition, we check whether at least one number was found. If a number exists, age_num[0] retrieves the value at index 0, which is the first number found. In this dataset, the first number represents the actual age, so we convert it to an integer and return it.

Note: The .apply() function takes each value from the Age column and passes it to the lambda function.